test/replication-luatest/gh_7581_downstream_lag_test.lua

```diff
@@ -0,0 +1,84 @@
+local fiber = require("fiber")
+local t = require("luatest")
+local server = require("test.luatest_helpers.server")
+local cluster = require("test.luatest_helpers.cluster")
+
+local g = t.group("downstream-lag-after-master-switch")
+
+local POLL_TIMEOUT = 0.1
+--
+-- Downstream lag calculations were off after master switch or after a period of
+-- time with no transactions: downstream lag showed time since the last
+-- processed transaction.
+--
+g.before_each(function(g)
+    g.cluster = cluster:new{}
+    local box_cfg = {
+        replication_timeout = POLL_TIMEOUT,
+        replication = {
+            server.build_instance_uri("server1"),
+            server.build_instance_uri("server2"),
+        },
+    }
+    g.server1 = g.cluster:build_and_add_server({
+        alias = "server1",
+        box_cfg = box_cfg,
+    })
+    g.server2 = g.cluster:build_and_add_server({
+        alias = "server2",
+        box_cfg = box_cfg,
+    })
+    g.cluster:start()
+    g.server1:exec(function()
+        box.schema.space.create("test")
+        box.space.test:create_index("pk")
+    end)
+end)
+
+g.after_each(function(g)
+    g.cluster:drop()
+end)
+
+local function get_downstream_lag(server)
+    return server:exec(function()
+        local other_id = 3 - box.info.id
```

Comment:
I don't like this line: there is a magic parameter there. I get that there are two servers with ids 1 and 2 and in sum it's 3. But it made me think for a moment. Maybe pass neighbour UUID, and get its ID from _cluster, return first box.info.replication[id].downstream.lag, where id ~= box.info.id.

Reply:
Yep, looks strange. Reworked like this. Also changed timeout in one place because saw 1 flaky failure in 256 runs:

<details><summary>diff</summary>

```diff
diff --git a/test/replication-luatest/gh_7581_downstream_lag_test.lua b/test/replication-luatest/gh_7581_downstream_lag_test.lua
index 543dc73f8..d4a89240e 100644
--- a/test/replication-luatest/gh_7581_downstream_lag_test.lua
+++ b/test/replication-luatest/gh_7581_downstream_lag_test.lua
@@ -6,6 +6,7 @@ local cluster = require("test.luatest_helpers.cluster")
 local g = t.group("downstream-lag-after-master-switch")
 
 local POLL_TIMEOUT = 0.1
+local DISCONNECT_TIMEOUT = 4 * POLL_TIMEOUT
 --
 -- Downstream lag calculations were off after master switch or after a period of
 -- time with no transactions: downstream lag showed time since the last
@@ -39,13 +40,13 @@ g.after_each(function(g)
     g.cluster:drop()
 end)
 
-local function get_downstream_lag(server)
-    return server:exec(function()
-        local other_id = 3 - box.info.id
+local function get_downstream_lag(master, replica)
+    local id = replica:instance_id()
+    return master:exec(function(id)
         return require("luatest").helpers.retrying({}, function()
-            return box.info.replication[other_id].downstream.lag
+            return box.info.replication[id].downstream.lag
         end)
-    end)
+    end, {id})
 end
 
 local function wait_downstream_updated(master, replica)
@@ -66,13 +67,13 @@ g.test_downstream_lag = function(g)
     end)
     g.server2:wait_vclock_of(g.server1)
     wait_downstream_updated(g.server1, g.server2)
-    local lag = get_downstream_lag(g.server1)
+    local lag = get_downstream_lag(g.server1, g.server2)
     t.assert(lag ~= 0, "Real lag value is updated")
-    t.assert(lag < POLL_TIMEOUT, "Lag value is sane")
+    t.assert(lag < DISCONNECT_TIMEOUT, "Lag value is sane")
     -- Let a couple of pings pass through.
     fiber.sleep(2 * POLL_TIMEOUT);
-    lag = get_downstream_lag(g.server1)
-    t.assert_equals(lag, get_downstream_lag(g.server1),
+    lag = get_downstream_lag(g.server1, g.server2)
+    t.assert_equals(lag, get_downstream_lag(g.server1, g.server2),
                     "Lag doesn't change on standby")
     g.server2:exec(function()
         box.space.test:insert{2}
@@ -81,7 +82,7 @@ g.test_downstream_lag = function(g)
     -- downstream lag.
     g.server1:wait_vclock_of(g.server2)
     wait_downstream_updated(g.server1, g.server2)
-    t.assert_equals(lag, get_downstream_lag(g.server1),
+    t.assert_equals(lag, get_downstream_lag(g.server1, g.server2),
                     "Lag doesn't change when there are updates from remote \
                      servers")
 end

```

</details>